airbyte-cdk/java/airbyte-cdk/s3-destinations/src/testFixtures/kotlin/io/airbyte/cdk/integrations/destination/s3/S3DestinationAcceptanceTest.kt

```diff
@@ -185,6 +203,304 @@ protected constructor(protected val outputFormat: FileUploadFormat) :
+                .withRecord(
+                    AirbyteRecordMessage()
+                        .withStream(configuredCatalog.streams[0].stream.name)
+                        .withEmittedAt(Instant.now().toEpochMilli())
+                        .withData(
+                            Jsons.jsonNode(
+                                ImmutableMap.builder<Any, Any>()
+                                    .put("id", 1)
+                                    .put("currency", "USD")
+                                    .put("date", "2020-03-31T00:00:00Z")
+                                    .put("HKD", 10.1)
+                                    .put("NZD", 700.1)
+                                    .build(),
+                            ),
+                        ),
+                ),
+            AirbyteMessage()
+                .withType(AirbyteMessage.Type.STATE)
+                .withState(
+                    AirbyteStateMessage()
+                        .withData(Jsons.jsonNode(ImmutableMap.of("checkpoint", 2))),
+                ),
+            AirbyteMessage()
+                .withType(AirbyteMessage.Type.TRACE)
+                .withTrace(
+                    AirbyteTraceMessage()
+                        .withType(AirbyteTraceMessage.Type.STREAM_STATUS)
+                        .withStreamStatus(
+                            AirbyteStreamStatusTraceMessage()
+                                .withStreamDescriptor(descriptor)
+                                .withStatus(streamStatus)
+                        ),
+                ),
+        )
+    }
+
+    private fun getSyncMessagesFixture2(): List<AirbyteMessage> {
+        return MoreResources.readResource(
+                DataArgumentsProvider.Companion.EXCHANGE_RATE_CONFIG.getMessageFileVersion(
+                    getProtocolVersion(),
+                ),
+            )
+            .trim()
+            .lines()
+            .map { Jsons.deserialize(it, AirbyteMessage::class.java) }
+    }
+
+    @Test
+    fun testOverwriteSyncPreRefreshAndPostSupport() {
+        if (!implementsOverwrite()) {
+            LOGGER.info { "Destination's spec.json does not support overwrite sync mode." }
+            return
+        }
+
+        // Run sync with some messages
+        val catalogPair =
+            getTestCatalog(SyncMode.FULL_REFRESH, DestinationSyncMode.OVERWRITE, 42, null, null)
+        val config = getConfig()
+        val firstSyncMessages =
+            getFirstSyncMessagesFixture1(
+                catalogPair.first,
+                AirbyteStreamStatusTraceMessage.AirbyteStreamStatus.COMPLETE,
+            )
+        runSyncAndVerifyStateOutput(config, firstSyncMessages, catalogPair.first, false)
+
+        // Change the generationId, we always assume platform sends a monotonically increasing
+        // number
+        val catalogPair2 =
+            getTestCatalog(SyncMode.FULL_REFRESH, DestinationSyncMode.OVERWRITE, 43, 0, 0)
```

Comment:
nit: named params for int literals

(... I think this actually incremented the sync ID, rather than the gen ID?)

Reply:
Or are we testing `null -> 0`? Ie, the "migration"? Though won't the first gen id be 1?

Reply:
aha. that makes sense. Would be good to clarify the comment here then (null -> 0 isn't intuitively a "monotonically increasing" sequence)

platform starts generations at 0, but for full refresh syncs, immediately increments them to 1 when you run a sync. So full refresh syncs never see gen 0.

(_incremental_ syncs can see gen 0, which is what bit us with the postgres incident)

Reply:
Yeah. I wasn't clear if the first gen id is 0 or 1. thanks @edgao  that clarifies what I was worried about. I'll switch overwrite to 1 and add a test for incremental with genId 0 (this test kinda proxies that but not representation of real scenario).